entries/ghatem-fpc/src/onebrc.pas

```diff
@@ -0,0 +1,274 @@
+// given a line of data, extract the station name and temperature, as strings.
+var
+  I: Int64;
+begin
+  // we're looking for the semicolon ';', but backwards since there's fewer characters to traverse
+  // a thermo measurement must be at least 3 characters long (one decimal, one period, and the units)
+  // e.g. input: Rock Hill;-54.3
+  // can safely skip 3:      ^^^
+  I := aEnd-3;
+
+  while True do begin
+    if FData[I] = ';' then
+      break;
+    Dec(I);
+  end;
+
+  // I is the position of the semi-colon, extract what's before and after it
+  SetString(aStation, pAnsiChar(@FData[aStart]), i-aStart);
+  SetString(aTempStr, pAnsiChar(@FData[i+1])   , aEnd-i);
+end;
+
+//---------------------------------------------------
+
+procedure TOneBRC.DoTempsDict (var aDict: TFPHashList; var aStrVal: AnsiString;
+                       var aIntVal: SmallInt);
+// the parsed temperatures are as strings, we need them as smallInts.
+// however, a few problems:
+// - StrToInt is VERY expensive
+// - Val is less expensive than StrToInt, but still very expensive
+// solution:
+// temperatures vary (between at most -100.0 and 100.0, that's 200*10=2000 possible different readings,
+// regardless if there are 100M or 1B lines.
+// practically in this input file, there are 1998 readings exactly:
+// convert each one once, and store the resulting smallInt in a dictionary
+{TODO: thread-safety}
+var vSuccess: Integer;
+    vIdx: Integer;
+    pInt: PSmallInt;
+begin
+  // replicate the last char, then drop it
+  aStrVal[Length(aStrVal) - 1] := aStrVal[Length(aStrVal)];
+  SetLength (aStrVal, Length(aStrVal) - 1);
+
+  //vIdx := aDict.FindIndexOf(aStrVal);
+  //if vIdx >= 0 then begin
+    //pInt := PSmallInt (aDict.Items[vIdx]);
+    //aIntVal := pInt^;
+  //end
+  //else begin
+    Val (aStrVal, aIntVal, vSuccess);
+    if vSuccess <> 0 then
+      raise Exception.Create('cannot decode value');
+    //FTempArr[FTempIter] := aIntVal;
+    //aDict.Add (aStrVal, @(FTempArr[FTempIter]));
+    //Inc (FTempIter);
+  //end;
+end;
+
+//---------------------------------------------------
+
+constructor TOneBRC.Create;
+begin
+  FStationsDict := TFPHashList.Create;
+  FStationsDict.Capacity := 45000;
+
+  FStations := TStringList.Create;
+  FStations.Capacity := 45000;
+  FStations.UseLocale := False;
+
+  FTempIter := 0;
+end;
+
+//---------------------------------------------------
+
+function TOneBRC.mORMotMMF (const afilename: string): Boolean;
+begin
+  Result := FMemoryMap.Map (aFilename);
+  if Result then begin
+    FData     := FMemoryMap.Buffer;
+    FDataSize := FMemoryMap.Size;
+  end;
+end;
+
+//---------------------------------------------------
+
+procedure TOneBRC.SingleThread;
+var
+  vTempDict: TFPHashList;
+  i: Int64;
+  vStation: AnsiString;
+  vTempStr: AnsiString;
+  vTemp: SmallInt;
+  vData: PStationData;
+  vLineStart: Int64;
+  vRecIdx: Integer;
+  vIdx: Integer;
+begin
+  // expecting 1998 different temperature measurements
```

Comment:
my guess is that there are 41343 stations in the benchmark data.

Reply:
@synopse you're right, there's ~40K stations.
here, I was trying to optimize on converting a string value (e.g. "-531") to a SmallInt value (-531), by reducing the number of calls to `Val (...)`.  Between -100.0 and 100.0 degrees Celcius, there are ~2000 possible string values for temperatures.

On Delphi, the TDictionary <AnsiString, SmallInt> was able to store and retrieve data much faster than executing `Val` 1B times.
However, on FPC, turns out TFPHashList is slower, probably due to the different hashing function used, so much so that executing `Val` 1B times is faster.

If I am able to improve the hashing algorithm of the hashmap (TFPHashList / TDictionary), caching the converted values might prove useful again.

Reply:
yes, I understand but the whole mapping idea is just wrong
Two advises:
- the overhead of using a temporary AnsiString is likely to be higher than parsing its value
- a dictionary will compute a hash when retrieving the value: manual parsing of the -xx.x to xx.x text is likely to be faster than this hash itself, and will pollute the CPU cache

In practice, parsing the parameter could be done with a few lines of code directly on the input text chars, with no temp string nor val() or StrToFloat().

Reply:
don't know how i didn't think about it before... a simple Ord to map the character to its numerical equivalent, multiplied by the decimal position, does the job and is SO much faster indeed.  No interim string, no Val() nor StrToFloat: one string extraction is out of the way, one to go!

I've been trying for the past few days to extract the station name without having to use an interim string, but rather point directly at the pointer data returned by the memory map.  Then again, without a ShortString as key, TFPHashList can't be used.  The Generics.Collection alternative in both `FPC` and `Delphi` modes is 2x slower than TFPHashList when mapping `string -> PStationData`.  Trying to game the hashing by using MD5 or SHA1 for example, is also much slower, probably because the length of the string, and string manipulation in general.

the docs at https://wiki.freepascal.org/hash also mention a CRC algorithm which returns a cardinal.  In Delphi, I would have used a TDictionary<Cardinal, PStationData>, but the Lazarus generics are relatively much slower.
Yet, this approach has yielded much better results than any of my prior results, and the actual string extraction (using `SetString`) is only needed once per station, to later on sort by name.

Any recommendation for a high-performance hashmap implementation in FPC (not necessarily threaded - still working on the 1-thread version) that would accept a Cardinal as key?
Of course, also open to other recommendations / readings you might have.

Cordialement,
Georges